test/test_type_promotion.py

```diff
@@ -0,0 +1,155 @@
+import torch
+import unittest
+
+from common_utils import TestCase, run_tests, load_tests
+
+# load_tests from common_utils is used to automatically filter tests for
+# sharding on sandcastle. This line silences flake warnings
+load_tests = load_tests
+
+class TestTypePromotion(TestCase):
+
+    def setUp(self):
+        super(TestTypePromotion, self).setUp()
+        torch.set_default_dtype(torch.float32)
+        self.device = 'cpu'
+
+    # In-place operations don't promote.
+    # int+float==float but int.add_(float)==int.
+    # Promoting inplace would require re-allocating and copying the memory of the
+    # tensor data, since element size could change.
+    def test_inplace(self):
+        int_tensor = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: int_tensor.add_(1.5))
+
+        expected = torch.ones([4, 4, 4], device=self.device)
+        expected = expected.to(torch.int32)
+
+        long_tensor = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        int_tensor.add_(long_tensor)
+        int_tensor.add_(1)
+        self.assertEqual(int_tensor, expected + 2)
+
+        byte_tensor = torch.ByteTensor([1, 1, 1])
```

Comment:
can you use new style constructors?  `torch.tensor([1, 1, 1], dtype=torch.uint8)`.

Reply:
sure, will do